skbio/diversity/beta/_fast_unifrac.py

```diff
@@ -0,0 +1,390 @@
+# ----------------------------------------------------------------------------
+
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+
+from skbio.util._decorator import experimental
+from ._unifrac import _validate
+from skbio.diversity._fast_base import counts_and_index
+from skbio.diversity._fast_base_cy import tip_distances
+
+
+def unifrac(m, i, j):
+    """Calculates unifrac(i,j) from m.
+
+    Parameters
+    ----------
+    m : np.array
+        A 1D vector that represents the lengths in the tree in postorder.
+    i,j : np.array
+        A slice of states from m in postorder.
+
+    Returns
+    -------
+    float
+        Unweighted unifrac metric
+
+    Notes
+    -----
+    This is cogent.maths.unifrac.fast_tree.unifrac, but there are
+    other metrics that can (should?) be ported like:
+     - unnormalized_unifrac
+     - G
+     - unnormalized_G
+    """
+    _or = np.logical_or(i, j),
+    _and = np.logical_and(i, j)
+    return 1 - ((m * _and).sum() / (m * _or).sum())
+
+
+def w_unifrac(m, i, j, i_sum, j_sum):
+    """Calculates weighted unifrac(i, j) from m
+
+    Parameters
+    ----------
+    m : np.array
+        A 1D vector that represents the lengths in the tree in postorder.
+    i, j : np.array
+        A slice of states from m in postorder.
+    i_sum, j_sum: float
+        Counts of the observations in each environment
+
+    Returns
+    -------
+    float
+        The weighted unifrac score
+    """
+    if i_sum:
+        i_ = i / i_sum
+    else:
+        i_ = 0.0
+
+    if j_sum:
+        j_ = j / j_sum
+    else:
+        j_ = 0.0
+
+    return (m * abs(i_ - j_)).sum()
+
+
+def make_pdist(counts, obs_ids, tree, indexed=None, metric=unifrac,
+               normalized=False, **kwargs):
+    """Construct a metric for use with skbio.diversity.beta.pw_distances
+
+    Parameters
+    ----------
+    counts : np.ndarray
+        A matrix of environment counts where each row is an environment, and
+        each column is an observation. The observations are expected to be in
+        index order w.r.t. obs_ids, but do not need to be in tree order.
+    obs_ids : np.ndarray
+        A vector of observation IDs. These IDs must map to tips in the tree.
+    tree : skbio.tree.TreeNode
+        A tree that represents the relationships between the observations.
+    indexed : dict, optiona;
+        The result of skbio.diversity.beta.index_tree
+    metric : function, {unweighted_unifrac_fast, weighted_unifrac_fast}
+        The specific metric to use.
+    normalized : bool
+        Whether the weighted unifrac calculation should be normalized.
+
+    Notes
+    -------
+    example usage
+
+    metric, counts, length = make_unweighted_pdist(input_counts, tip_ids, tree)
+    mat = pw_distances(metric, counts, ids=['%d' % i for i in range(10)])
+    """
+    count_array, indexed = counts_and_index(counts, obs_ids, tree, indexed)
```

Comment:
just a note, it would be nice to be able to cache `count_array` as often you end up computing both unweighted and weighted -- no point in eating this expense twice